docs/ai/conceptual/understanding-tokens.md

```diff
@@ -0,0 +1,121 @@
+
+Tokens are the words or character sets that large language models (LLMs) decompose text into. During training, an LLM analyzes the semantic relationships between these tokens, such as how commonly they're used together or if they're used in similar contexts. Once trained, an LLM uses those patterns and relationships to generate a sequence of output tokens based on the input sequence.
+
+## What are tokens?
+
+The first step when training an LLM or inputting a prompt is tokenization. The text is broken down into a set of strings, known as _tokens_. These tokens can be distinct words, character sets, or combinations of words and punctuation. The set of unique tokens a model has been trained on is referred to as its _vocabulary_.
+
+For example, consider the following sentence:
+
+> I heard a dog bark loudly at a cat
+
+This text could be tokenized as:
+
+- I
+- heard
+- a
+- dog
+- bark
+- loudly
+- at
+- a
+- cat
+
+Only the first occurrence of the "a" token will be added to the model's vocabulary. All other occurrences can be represented by that first token.
+
+With a sufficiently large set of training text, a vocabulary of many thousands of tokens can be compiled.
+
+## How are tokens determined?
+
+The specific tokenization method varies by LLM. Some common tokenization methods include:
+
+- Word tokenization (text is split into individual words based on a delimiter)
+- Character tokenization (text is split into individual characters)
+- Subword tokenization (text is split into partial words or character sets)
+
+For example, the GPT models, developed by OpenAI, use _Byte-Pair Encoding_ (BPE) which is a type of subword tokenization. OpenAI provides [a tool to visualize how text will be tokenized](https://platform.openai.com/tokenizer).
+
+There are benefits and disadvantages to each tokenization method:
+
+**Smaller tokens (character or subword tokenization)**
+
+- _Pros_:
+  - Allows the model to handle a wider range of inputs, such as unknown words, typos, or complex syntax
+  - Might allow the vocabulary size to be reduced, requiring less memory resources
+- _Cons_:
+  - A given text is broken into more tokens, requiring additional computational resources while processing
+  - Given a fixed token limit, the maximum size of the model's input and output is smaller
+
+**Larger tokens (word tokenization)**
+
+- _Pros_:
+  - A given text is broken into fewer tokens, requiring less computational resources while processing
+  - Given the same token limit, the maximum size of the model's input and output is larger
+- _Cons_:
+  - Might cause an increased vocabulary size, requiring more memory resources
+  - Can limit the models ability to handle unknown words, typos, or complex syntax
+
+## How tokens are used
+
+While training, after a text has been tokenized, each unique token is assigned an ID.
+
+Consider the following sentence:
+
+> I heard a dog bark loudly at a cat
+
+After tokenization, the following IDs could be assigned:
```

Comment:
"After using a word tokenization method, a model could assign token IDs like this:"